### PR TITLE
Allow "Semesters to Show" field to be temporarily blank

### LINKED ISSUE #76: [Bug] Cannot backspace on "Semesters to Show" field in Aggregate FCEs when the input is a single digit
### Expected Behavior

A user should be able to clear the entire field (e.g., from the default value of 2) and key in a new number.

Instead, emptying the field keeps the current value. This makes it difficult to change a single-digit input (e.g. from 2 to 3).

### Steps to Reproduce

1. Go to [https://cmucourses.com/](cmucourses.com)
2. Change the "Semesters to Show" field/

### Device Information

Desktop/Mobile: Desktop and Mobile
OS: Found on Linux (EndeavourOS) 5.19.0 / Windows 11 / Android 12
Browser Version: Firefox / Chrome

### PR DESCRIPTION
**Pull request by Lhwhatever**
Should fix #76.

This allows backspacing to work as a user might expect; i.e. they can clear the field and enter a new value. If the user leaves the field blank and unfocuses the field, it will revert back to its last valid value.

This was more complicated to implement than expected. An auxiliary state has to be implemented to handle the case when the field is temporarily blank (a future developer may be able to address this). This auxiliary state updates the global state (i.e. calls the dispatch function) through a useEffect hook.

In addition, due to the limitations of the input[type=number] element, we need to add a onKeyPress handler to disable key presses for non-numeric keys, as the onChange handler cannot detect when non-numeric input is given to the field.

# Description

Closes #76 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

**Test Configuration**:

- Node.js version: 18.9.0
- Desktop/Mobile: Desktop
- OS: Linux 5.19.0 (EndeavourOS)
- Browser: Firefox 105.0.1

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: frontend/src/components/Aggregate.tsx [hunk 1]
```diff
@@ -4,6 +4,9 @@ import { useAppDispatch, useAppSelector } from "../app/hooks";
 import { userSlice } from "../app/user";
 import { Semester } from "../app/types";
 
+const numeric = /\d/;
```
Comment (btwj) [inline frontend/src/components/Aggregate.tsx:7]: For consistency, name this `numericRegex` too.

File: frontend/src/components/Aggregate.tsx [hunk 2]
```diff
@@ -18,6 +21,42 @@ const Aggregate = () => {
...
+    }
+  };
+
+  // Restore the last valid value if the field is left blank and unfocused
+  const handleNumSemsFieldBlur: React.FocusEventHandler<
+    HTMLInputElement
+  > = () => {
+    if (numSemsField === undefined) setNumSemsField(numSemesters);
+  };
+
```
Comment (btwj) [inline frontend/src/components/Aggregate.tsx:59]: I don't think there's a way to get around needing an auxiliary state, so this is fine. You might be able to streamline it though: the input has `numSemsField`. Once the input is unfocused, we either restore to the redux state if its value is nonnumeric, or update the redux state if its value is a valid number.

That way, we don't need to override the keyDown event. I tested it locally and currently, there are some odd behaviors where after backspacing, I can't type any numbers anymore. I think the above method would be more robust.

File: frontend/src/components/Aggregate.tsx [hunk 3]
```diff
@@ -28,11 +67,11 @@ const Aggregate = () => {
...
-            value={numSemesters}
+            value={numSemsField === undefined ? "" : numSemsField}
             className="border-gray-200 min-w-0 flex-auto rounded border px-2 py-1 text-sm bg-transparent"
-            onChange={(e) => {
-              setNumSemesters(parseInt(e.target.value));
-            }}
+            onChange={handleNumSemsFieldChange}
+            onBlur={handleNumSemsFieldBlur}
+            onKeyPress={handleNumSemsFieldKeyDown}
           />
```
Comment (btwj) [inline frontend/src/components/Aggregate.tsx:75]: Also, could you refactor the numSems input out as another component in the same file? The logic is getting a bit complicated.

**Review by btwj — CHANGES_REQUESTED** [review]
Some comments on the current behavior and refactoring.

**Review by Lhwhatever — COMMENTED** [review]
Resolved in 810383e24f03043268c416e0a53592bf1408cab8 and cbd4e45219a2ac7b201fbe55bfdb721f38757973

**Pull request merged**